### DISCUSSION TITLE
accept before syn

### DISCUSSION THREAD

**Discussion by estherjsong** [ANLAB-KAIST/KENSv3#184] [Project Q&A 2. Transport Layer (TCP)]
in AfterAccept testcase, test tries to accept and close before the handshake is established.
And it makes problem because (1) if we close due to no pending queue, it keeps sending syn, and (2) if we leave it open and send synack, it keeps sending finack.
what is the correct behavior of server in this situation?

(1)
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/40757981/d55ca72f-f329-4d46-8cb8-05df9ee1b7d5)

(2)
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/40757981/d1a75420-98f9-498d-aea7-f90e69a8feb7)

![image](https://github.com/ANLAB-KAIST/KENSv3/assets/40757981/a0617fbf-e682-49f6-8f11-66ba4a0b2f64)

So, in the last picture, 
3:0 means 'sockfd : pending queue size'
and accept backlog EGAIN is occured because the pending queue size is 0.
and then it tries close (it printes 'closing' in the log)
then, in the last part 1:0 means the queue size become 1 because test sends syn to server.

**Comment by Pusnow** [comment]
> test tries to accept and close before the handshake is established.

This cannot happen. The `accept()` call only return fd for an established connection, so it is impossible close unestablished connection. If there is no established connection, `accept()` is blocked.

> And it makes problem because (1) if we close due to no pending queue, it keeps sending syn, and (2) if we leave it open and send synack, it keeps sending finack.

Regardless of `accept()` calls, the server has to respond to the client. SYN-ACK for SYN, ACK for FIN-ACK.

> and accept backlog EGAIN is occured because the pending queue size is 0.

This is a non-blocking behavior of `accept()`. In KENS, we only implement blocking `accept()`, so you should block the `accept()` call rather than returning EAGAIN.
